Security/Authentication/Provider/AuthenticationProviderDecorator.php

```diff
@@ -0,0 +1,49 @@
+<?php
+namespace Scheb\TwoFactorBundle\Security\Authentication\Provider;
+
+use Scheb\TwoFactorBundle\Security\Authentication\Token\TwoFactorToken;
+use Symfony\Component\Security\Core\Authentication\Provider\AuthenticationProviderInterface;
+use Symfony\Component\Security\Core\Authentication\Token\AnonymousToken;
+use Symfony\Component\Security\Core\Authentication\Token\TokenInterface;
+
+class AuthenticationProviderDecorator implements AuthenticationProviderInterface
```

Comment:
This is the class that is wrapped around every authentication provider, to "magically" transform the authenticated token into a un-authenticated `TwoFactorToken`, which hides away the authenticated token.